File: domain/src/main/java/com/github/anrimian/musicplayer/domain/controllers/MusicPlayerController.java
package com.github.anrimian.musicplayer.domain.controllers;

import com.github.anrimian.musicplayer.domain.models.composition.source.CompositionSource;
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance;
import com.github.anrimian.musicplayer.domain.models.player.events.PlayerEvent;

import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;

/**
 * Created on 10.11.2017.
 */

public interface MusicPlayerController {

    Observable<PlayerEvent> getEventsObservable();

    void prepareToPlay(CompositionSource composition);

    void stop();

    void resume();

    void resume(int delay);

    void pause();

    void seekTo(long position);

    void setVolume(float volume);

    void setSoundBalance(SoundBalance soundBalance);

    Observable<Long> getTrackPositionObservable();

    void seekBy(long millis);

    Single<Long> getTrackPosition();

    void setPlaybackSpeed(float speed);

    Observable<Float> getCurrentPlaybackSpeedObservable();

    Observable<Boolean> getSpeedChangeAvailableObservable();
}

File: domain/src/main/java/com/github/anrimian/musicplayer/domain/controllers/MusicPlayerController.kt
package com.github.anrimian.musicplayer.domain.controllers

import com.github.anrimian.musicplayer.domain.models.composition.content.CompositionContentSource
import com.github.anrimian.musicplayer.domain.models.player.SoundBalance
import com.github.anrimian.musicplayer.domain.models.player.events.MediaPlayerEvent
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.core.Single

interface MusicPlayerController {

    fun prepareToPlay(source: CompositionContentSource): Completable

    fun stop()

    fun pause()

    fun resume()

    fun seekTo(position: Long)

    fun setVolume(volume: Float)

    fun setSoundBalance(soundBalance: SoundBalance)

    fun getTrackPosition(): Single<Long>

    fun getDuration(): Single<Long>

    fun setPlaybackSpeed(speed: Float)

    fun getTrackPositionObservable(): Observable<Long>

    fun getPlayerEventsObservable(): Observable<MediaPlayerEvent>

    fun getCurrentPlaybackSpeedObservable(): Observable<Float>

    fun getSpeedChangeAvailableObservable(): Observable<Boolean>

}